src/main/java/org/spongepowered/common/mixin/core/world/MixinWorld.java

```diff
@@ -563,21 +585,63 @@ public void updateWorldGenerator() {
     }
 
     @Override
-    public SpongeChunkGenerator createChunkProvider(SpongeWorldGenerator newGenerator) {
+    public SpongeChunkGenerator createChunkGenerator(SpongeWorldGenerator newGenerator) {
         return new SpongeChunkGenerator((net.minecraft.world.World) (Object) this, newGenerator.getBaseGenerationPopulator(),
                 newGenerator.getBiomeGenerator());
     }
 
     @Override
+    public SpongeWorldGenerator createWorldGenerator(DataContainer settings) {
+        // Minecraft uses a string for world generator settings
+        // This string can be a JSON string, or be a string of a custom format
+
+        // Try to convert to custom format
+        Optional<String> optCustomSettings = settings.getString(DataQueries.WORLD_CUSTOM_SETTINGS);
+        if (optCustomSettings.isPresent()) {
+            return this.createWorldGenerator(optCustomSettings.get());
+        }
+
+        final StringWriter writer = new StringWriter();
+        try {
+            HoconConfigurationLoader.builder().setRenderOptions(ConfigRenderOptions.concise().setJson(true))
+                    .setSink(() -> new BufferedWriter(writer)).build().save(ConfigurateTranslator.instance().translateData(settings));
+        } catch (Exception e) {
+            SpongeImpl.getLogger().warn("Failed to convert settings from [{}] for GeneratorType [{}] used by World [{}].", settings, ((net
+                    .minecraft.world.World) (Object) this).getWorldType(), this, e);
+        }
+
+        return this.createWorldGenerator(writer.toString());
+    }
+
+    @Override
+    public SpongeWorldGenerator createWorldGenerator(String settings) {
+        final WorldServer worldServer = (WorldServer) (Object) this;
+        final WorldType worldType = worldServer.getWorldType();
+        final IChunkGenerator chunkGenerator;
+        final BiomeProvider biomeProvider;
+        if (worldType instanceof SpongeWorldType) {
+            chunkGenerator = ((SpongeWorldType) worldType).getChunkGenerator(worldServer, settings);
+            biomeProvider = ((SpongeWorldType) worldType).getBiomeProvider(worldServer);
+        } else {
+            final WorldProvider worldProvider = worldServer.provider;
+            ((IMixinWorldProvider) worldProvider).setGeneratorSettings(settings);
+            chunkGenerator = worldProvider.createChunkGenerator();
+            biomeProvider = worldServer.provider.biomeProvider;
```

Comment:
TODO @simon816 's change on master to call getBiomeProvider is more correct here. Do that.